File: core/src/main/java/fr/sncf/osrd/api/pathfinding_constraints/LoadingGaugeConstraints.java
package fr.sncf.osrd.api.pathfinding_constraints;

import fr.sncf.osrd.infra.api.reservation.ReservationRoute;
import fr.sncf.osrd.infra.api.signaling.SignalingRoute;
import fr.sncf.osrd.train.RollingStock;
import fr.sncf.osrd.utils.graph.Pathfinding;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Function;

public record LoadingGaugeConstraints(
        Collection<RollingStock> rollingStocks
) implements Function<SignalingRoute, Collection<Pathfinding.Range>> {

    @Override
    public Collection<Pathfinding.Range> apply(SignalingRoute reservationRoute) {
        var res = new HashSet<Pathfinding.Range>();
        for (var stock : rollingStocks)
            res.addAll(getBlockedRanges(stock, reservationRoute.getInfraRoute()));
        return res;
    }

    /**
     * Returns the sections of the given route that can't be used by the given rolling stock
     */
    private static Set<Pathfinding.Range> getBlockedRanges(RollingStock stock, ReservationRoute route) {
        var offset = 0;
        var res = new HashSet<Pathfinding.Range>();
        for (var trackRange : route.getTrackRanges()) {
            for (var entry : trackRange.getBlockedGaugeTypes().asMapOfRanges().entrySet()) {
                if (!entry.getValue().isCompatibleWith(stock.loadingGaugeType))
                    res.add(new Pathfinding.Range(
                            offset + entry.getKey().lowerEndpoint(),
                            offset + entry.getKey().upperEndpoint()
                    ));
            }
            offset += trackRange.getLength();
        }
        return res;
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/pathfinding/constraints/LoadingGaugeConstraints.kt
package fr.sncf.osrd.pathfinding.constraints

import fr.sncf.osrd.graph.PathfindingConstraint
import fr.sncf.osrd.path.implementations.buildTrainPathFromBlock
import fr.sncf.osrd.path.interfaces.TrainPath
import fr.sncf.osrd.railjson.schema.rollingstock.RJSLoadingGaugeType
import fr.sncf.osrd.sim_infra.api.*
import fr.sncf.osrd.utils.units.Offset
import fr.sncf.osrd.utils.units.OffsetRange

data class LoadingGaugeConstraints(
    val blockInfra: BlockInfra,
    val infra: RawSignalingInfra,
    val loadingGaugeType: RJSLoadingGaugeType,
) : PathfindingConstraint {
    override fun apply(edge: BlockId): Collection<OffsetRange<Block>> {
        val res = HashSet<OffsetRange<Block>>()
        val path = buildTrainPathFromBlock(infra, blockInfra, edge)
        res.addAll(getBlockedRanges(loadingGaugeType, path))
        return res
    }

    /** Returns the sections of the given block that can't be used by the given rolling stock */
    private fun getBlockedRanges(
        type: RJSLoadingGaugeType,
        path: TrainPath,
    ): Collection<OffsetRange<Block>> {
        return path
            .getLoadingGauge()
            .toSet()
            .filter { !it.value.isCompatibleWith(LoadingGaugeTypeId(type.ordinal.toUInt())) }
            .map { (lower, upper) -> OffsetRange(Offset(lower), Offset(upper)) }
    }
}
